### PR TITLE
380 enforce pep 8

### PR DESCRIPTION
**Pull request by bhagemeier**
## Description

This set of changes implements a mechanism for developers to automatically do style checking and PEP-8 fixes. To this end, a [pre-commit](https://pre-commit.com/) hook is employed that is configured to run flake8 and the [Black code formatter](https://github.com/psf/black).

A lot of the issues reported have already been fixed and committed individually. Some remaining ones will have to be dealt with later, such as [E501](https://lintlyci.github.io/Flake8Rules/rules/E501.html), [F403](https://lintlyci.github.io/Flake8Rules/rules/F403.html), [F401](https://lintlyci.github.io/Flake8Rules/rules/F401.html). The latter two are exclusively in __init__.py files, so it's a matter of debate, whether this is really a problem.

Fixes: #380 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

Have you handled and tested all split configurations?
- n/a

### CODE REVIEW

File: heat/core/arithmetics.py [hunk 1]
```diff
@@ -107,32 +107,51 @@ def diff(a, n=1, axis=-1):
     size = a.comm.size
     rank = a.comm.rank
     ret = a.copy()
-    for _ in range(n):  # work loop, runs n times. using the result at the end of the loop as the starting values for each loop
+    for _ in range(
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: This seems wrong. Comment should be above

File: heat/core/arithmetics.py [hunk 2]
```diff
@@ -107,32 +107,51 @@ def diff(a, n=1, axis=-1):
...
+        n
+    ):  # work loop, runs n times. using the result at the end of the loop as the starting values for each loop
         axis_slice = [slice(None)] * len(ret.shape)
         axis_slice[axis] = slice(1, None, None)
         axis_slice_end = [slice(None)] * len(ret.shape)
         axis_slice_end[axis] = slice(None, -1, None)
 
         arb_slice = [slice(None)] * len(a.shape)
-        arb_slice[axis] = 0  # build the slice for the first element on the specified axis
+        arb_slice[
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: This also seems wrong, comment should be above

File: heat/core/arithmetics.py [hunk 3]
```diff
@@ -107,32 +107,51 @@ def diff(a, n=1, axis=-1):
...
+            axis
+        ] = 0  # build the slice for the first element on the specified axis
         if rank > 0:
-            snd = ret.comm.Isend(ret.lloc[arb_slice].clone(), dest=rank - 1, tag=rank)  # send the first element of the array to rank - 1
-
-        dif = ret.lloc[axis_slice] - ret.lloc[axis_slice_end]  # standard logic for the diff with the next element
-        diff_slice = [slice(x) for x in dif.shape]  # need to slice out to select the proper elements of out
+            snd = ret.comm.Isend(
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: This also seems wrong, comment should be above

File: heat/core/arithmetics.py [hunk 4]
```diff
@@ -107,32 +107,51 @@ def diff(a, n=1, axis=-1):
...
+        ]  # need to slice out to select the proper elements of out
         ret.lloc[diff_slice] = dif
 
         if rank > 0:
             snd.wait()  # wait for the send to finish
         if rank < size - 1:
             cr_slice = [slice(None)] * len(a.shape)
-            cr_slice[axis] = 1  # slice of 1 element in the selected axis for the shape creation
-            recv_data = torch.ones(ret.lloc[cr_slice].shape, dtype=ret.dtype.torch_type())
+            cr_slice[
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: This also seems wrong, comment should be above

File: heat/core/arithmetics.py [hunk 5]
```diff
@@ -107,32 +107,51 @@ def diff(a, n=1, axis=-1):
...
+            ] = 1  # slice of 1 element in the selected axis for the shape creation
+            recv_data = torch.ones(
+                ret.lloc[cr_slice].shape, dtype=ret.dtype.torch_type()
+            )
             rec = ret.comm.Irecv(recv_data, source=rank + 1, tag=rank + 1)
             axis_slice_end = [slice(None)] * len(a.shape)
-            axis_slice_end[axis] = slice(-1, None)  # select the last elements in the selected axis
+            axis_slice_end[axis] = slice(
+                -1, None
+            )  # select the last elements in the selected axis
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: This also seems wrong, comment should be above

File: heat/core/arithmetics.py [hunk 6]
```diff
@@ -107,32 +107,51 @@ def diff(a, n=1, axis=-1):
...
             rec.wait()
-            ret.lloc[axis_slice_end] = recv_data.reshape(ret.lloc[axis_slice_end].shape) - ret.lloc[axis_slice_end]  # diff logic
+            ret.lloc[axis_slice_end] = (
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: This also seems wrong, comment should be above

File: heat/core/communication.py [hunk 7]
```diff
@@ -430,51 +465,71 @@ def __reduce_like(self, func, sendbuf, recvbuf, *args, **kwargs):
         # contiguous or differently strided, they have to be made matching (if possible) first.
         if isinstance(sendbuf, torch.Tensor):
             # convert the send buffer to a pointer, number of elements and type are identical to the receive buffer
-            dummy = sendbuf.contiguous()  # make a contiguous copy and reassign the storage, old will be collected
-            sendbuf.set_(dummy.storage(), dummy.storage_offset(), size=dummy.shape, stride=dummy.stride())
+            dummy = (
+                sendbuf.contiguous()
+            )  # make a contiguous copy and reassign the storage, old will be collected
```
Comment (Markus-Goetz) [inline heat/core/communication.py]: This also seems wrong, comment should be above

File: heat/core/dndarray.py [hunk 8]
```diff
@@ -1339,7 +1436,9 @@ def __getitem__(self, key):
 
             arr = torch.Tensor()
 
-            if isinstance(key, int):  # if a sigular index is given and the tensor is split
+            if isinstance(
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: This also seems wrong, comment should be above

File: heat/core/dndarray.py [hunk 9]
```diff
@@ -1351,7 +1450,9 @@ def __getitem__(self, key):
                 else:
                     new_split = self.split
 
-                if key in range(chunk_start, chunk_end) and self.split == 0:  # only need to adjust the key if split==0
+                if (
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: This also seems wrong, comment should be above

File: heat/core/dndarray.py [hunk 10]
```diff
@@ -1361,35 +1462,53 @@ def __getitem__(self, key):
...
+                            self.comm.rank
+                        ),
+                        ResourceWarning,
+                    )
                     # arr is empty and gout is zeros
 
-            elif isinstance(key, (tuple, list)):  # multi-argument gets are passed as tuples by python
+            elif isinstance(
+                key, (tuple, list)
+            ):  # multi-argument gets are passed as tuples by python
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: This also seems wrong, comment should be above

File: heat/core/dndarray.py [hunk 11]
```diff
@@ -1361,35 +1462,53 @@ def __getitem__(self, key):
...
 
                 if self.split >= len(gout):
                     new_split = len(gout) - 1 if len(gout) - 1 > 0 else 0
                 else:
                     new_split = self.split
 
-                if isinstance(key[self.split], slice):  # if a slice is given in the split direction
+                if isinstance(
+                    key[self.split], slice
+                ):  # if a slice is given in the split direction
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: This also seems wrong, comment should be above

File: heat/core/dndarray.py [hunk 12]
```diff
@@ -1361,35 +1462,53 @@ def __getitem__(self, key):
...
-                                        key[self.split].step if key[self.split].step else 1))
+                    key_set = set(
+                        range(
+                            key[self.split].start
+                            if key[self.split].start is not None
+                            else 0,
+                            key_stop
+                            if key_stop is not None
+                            else self.gshape[self.split],
+                            key[self.split].step if key[self.split].step else 1,
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: This seems very inconsistent. The last ternary expression is single-lines, while the two before are multi-line

File: heat/core/dndarray.py [hunk 13]
```diff
@@ -1400,18 +1519,26 @@ def __getitem__(self, key):
...
+                        ),
+                        ResourceWarning,
+                    )
                     # arr is empty
                     # gout is all 0s and is the proper shape
 
-            elif isinstance(key, slice) and self.split == 0:  # if the given axes are only a slice
+            elif (
+                isinstance(key, slice) and self.split == 0
+            ):  # if the given axes are only a slice
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: This also seems wrong, comment should be above

File: heat/core/linalg.py [hunk 14]
```diff
@@ -496,40 +617,70 @@ def c_block_setter(b_proc, a_proc, a_data, b_data):
...
+                    c_block_setter(
+                        a_proc=pr,
+                        b_proc=a.comm.rank,
+                        a_data=a_lp_data[pr],
+                        b_data=b._DNDarray__array,
+                    )
                     # check if there is a remainder on b on the last node (there shouldnt be)
                     if pr in a_rem_locs1:
-                        a_rem[:, pr] = a_lp_data[pr][:, -1]  # this is to save the data from B required by the remainders from dim1 of A
+                        a_rem[:, pr] = a_lp_data[pr][
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: This also seems wrong, comment should be above

File: heat/core/linalg.py [hunk 15]
```diff
@@ -496,40 +617,70 @@ def c_block_setter(b_proc, a_proc, a_data, b_data):
...
+                            :, -1
+                        ]  # this is to save the data from B required by the remainders from dim1 of A
 
-                    if b_rem_locs1.nelement() != 0:  # this loop is to take care of the remainders in the 0th dimension of A
+                    if (
+                        b_rem_locs1.nelement() != 0
+                    ):  # this loop is to take care of the remainders in the 0th dimension of A
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: Comment is wrong. If is no loop

This also seems wrong, comment should be above

File: heat/core/linalg.py [hunk 16]
```diff
@@ -564,21 +720,34 @@ def c_block_setter(b_proc, a_proc, a_data, b_data):
...
+            )
             return c
 
         elif split_10_flag:
             # for this case, only a sum is needed at the end
             a_rem_locs1 = (rem_map[:, 0, 1] == 1).nonzero()
-            b_rem_locs0 = (rem_map[:, 1, 0] == 1).nonzero()  # locations of the remainders in b
+            b_rem_locs0 = (
+                rem_map[:, 1, 0] == 1
+            ).nonzero()  # locations of the remainders in b
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: This also seems wrong, comment should be above

File: heat/core/linalg.py [hunk 17]
```diff
@@ -564,21 +720,34 @@ def c_block_setter(b_proc, a_proc, a_data, b_data):
...
-                res += a._DNDarray__array[:mB, i * kB:i*kB + kB] @ b._DNDarray__array[i * kB:i*kB + kB, :nB]
+                res += (
+                    a._DNDarray__array[:mB, i * kB : i * kB + kB]
+                    @ b._DNDarray__array[i * kB : i * kB + kB, :nB]
+                )
             if a.comm.rank in a_rem_locs1 and b.comm.rank in b_rem_locs0:
-                res += a._DNDarray__array[:, -1, None] @ b._DNDarray__array[None, -1, :]  # these Nones are used to change the dims
+                res += (
+                    a._DNDarray__array[:, -1, None] @ b._DNDarray__array[None, -1, :]
+                )  # these Nones are used to change the dims
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: This also seems wrong, comment should be above

File: heat/core/statistics.py [hunk 18]
```diff
@@ -498,13 +529,15 @@ def local_max(*args, **kwargs):
...
+        x, local_max, MPI.MAX, axis=axis, out=out, keepdim=keepdim
+    )
 
 
 def maximum(x1, x2, out=None):
     """
-    Compares two tensors and returns a new tensor containing the element-wise maxima. 
-    If one of the elements being compared is a NaN, then that element is returned. TODO: Check this: If both elements are NaNs then the first is returned. 
+    Compares two tensors and returns a new tensor containing the element-wise maxima.
+    If one of the elements being compared is a NaN, then that element is returned. TODO: Check this: If both elements are NaNs then the first is returned.
```
Comment (Markus-Goetz) [inline heat/core/statistics.py]: Seems like an overlong line. Shouldnt this be auto-moved to the next line?

Comment (bhagemeier) [reply heat/core/statistics.py]: In terms of line length, Black uncompromisingly does not wrap at least the doc strings. Other comments are wrapped, AFAIR. Here, it only removed trailing white space.

**Comment by bhagemeier** [comment]
There seems to be a pattern with Black. It seems to move comments to the wrong places when splitting lines. I had observed this for comments after the last of multiple statements per line, which ended up at the end of the line for the last of these statements after splitting. I guess, we'll have to fix the misplaced comments manually. I do not expect Black to move them again.

File: heat/core/dndarray.py (hunk 9 above)
Comment (bhagemeier) [reply heat/core/dndarray.py]: While the comments should certainly not be put in the way that Black did automatically, one can argue in the case of branch statements that one should put the explanation of the condition *inside* the corresponding branch body.

File: heat/core/statistics.py [hunk 19]
```diff
@@ -304,17 +323,20 @@ def average(x, axis=None, weights=None, returned=False):
         if x.gshape != weights.gshape:
             if axis is None:
                 raise TypeError(
-                    "Axis must be specified when shapes of x and weights "
-                    "differ.")
+                    "Axis must be specified when shapes of x and weights " "differ."
```
Comment (TheSlimvReal) [inline heat/core/statistics.py]: The quotation marks in front of differ can be removed

File: heat/core/tests/test_arithmetics.py [hunk 20]
```diff
@@ -56,8 +49,12 @@ def test_diff(self):
             arb_slice[dim] = slice(None)
             for ax in range(dim + 1):  # loop over the possible axis values
                 for sp in range(dim + 1):  # loop over the possible split values
-                    for nl in range(1, 4):  # loop to 3 for the number of times to do the diff
-                        lp_array = ht.manipulations.resplit(ht_array[arb_slice], sp)  # only generating the number once and then
+                    for nl in range(
+                        1, 4
```
Comment (TheSlimvReal) [inline heat/core/tests/test_arithmetics.py]: This also seems wrong, comment should be above

File: heat/core/tests/test_arithmetics.py [hunk 21]
```diff
@@ -56,8 +49,12 @@ def test_diff(self):
...
+                    ):  # loop to 3 for the number of times to do the diff
+                        lp_array = ht.manipulations.resplit(
+                            ht_array[arb_slice], sp
```
Comment (TheSlimvReal) [inline heat/core/tests/test_arithmetics.py]: This also seems wrong, comment should be above

File: heat/core/tests/test_arithmetics.py [hunk 22]
```diff
@@ -16,25 +16,18 @@ def setUpClass(cls):
         cls.a_vector = ht.float32([2, 2])
```
Comment (TheSlimvReal) [inline heat/core/tests/test_arithmetics.py]: should the linter even run on the test files?

File: heat/core/tests/test_random.py [hunk 23]
```diff
@@ -78,7 +78,9 @@ def test_rand(self):
         a = ht.random.rand(11, 15, 3, 7, split=2, comm=ht.MPI_WORLD)
         a = a.numpy()
         _, counts = np.unique(a, return_counts=True)
-        self.assertTrue((counts == 1).all())    # Assert that no value appears more than once
+        self.assertTrue(
+            (counts == 1).all()
```
Comment (TheSlimvReal) [inline heat/core/tests/test_random.py]: This also seems wrong, comment should be above

File: heat/core/tests/test_statistics.py [hunk 24]
```diff
@@ -817,26 +884,32 @@ def test_var(self):
...
             hold = list(range(len(dimensions)))
             hold.append(None)
             for i in hold:  # loop over the number of dimensions of the test array
                 z = ht.ones(dimensions, split=i)
                 res = z.var()
                 total_dims_list = list(z.shape)
                 self.assertTrue((res == 0).all())
-                for it in range(len(z.shape)):  # loop over the different single dimensions for mean
+                for it in range(
+                    len(z.shape)
```
Comment (TheSlimvReal) [inline heat/core/tests/test_statistics.py]: This also seems wrong, comment should be above

File: heat/core/types.py [hunk 25]
```diff
@@ -347,55 +348,57 @@ def heat_type_of(obj):
...
     # bool  uint8  int8   int16  int32  int64  float32 float64
-    [True,  True,  True,  True,  True,  True,  True,   True],  # bool
-    [False, True,  False, True,  True,  True,  True,   True],  # uint8
-    [False, False, True,  True,  True,  True,  True,   True],  # int8
-    [False, False, False, True,  True,  True,  True,   True],  # int16
-    [False, False, False, False, True,  True,  False,  True],  # int32
-    [False, False, False, False, False, True,  False,  True],  # int64
-    [False, False, False, False, False, False, True,   True],  # float32
-    [False, False, False, False, False, False, False,  True]   # float64
+    [True, True, True, True, True, True, True, True],  # bool
```
Comment (TheSlimvReal) [inline heat/core/types.py]: For more readability I would prefer to not change this formatting

File: heat/core/types.py [hunk 26]
```diff
@@ -347,55 +348,57 @@ def heat_type_of(obj):
...
     # bool  uint8  int8   int16  int32  int64  float32 float64
-    [True,  False, False, False, False, False, False,  False],  # bool
-    [False, True,  True,  True,  True,  True,  False,  False],  # uint8
-    [False, True,  True,  True,  True,  True,  False,  False],  # int8
-    [False, True,  True,  True,  True,  True,  False,  False],  # int16
-    [False, True,  True,  True,  True,  True,  False,  False],  # int32
-    [False, True,  True,  True,  True,  True,  False,  False],  # int64
-    [False, False, False, False, False, False, True,   True],   # float32
-    [False, False, False, False, False, False, True,   True]    # float64
+    [True, False, False, False, False, False, False, False],  # bool
```
Comment (TheSlimvReal) [inline heat/core/types.py]: For more readability I would prefer to not change this formatting

File: heat/core/dndarray.py (hunk 12 above)
Comment (bhagemeier) [reply heat/core/dndarray.py]: It probably has to do with [How Black wraps lines](https://github.com/psf/black#how-black-wraps-lines).

File: heat/ml/regression/lasso/lasso.py [hunk 27]
```diff
@@ -347,124 +354,124 @@ class PytorchLasso():
...
+        """
         # Get number of model parameters
         _, n = X.shape
 
-        # Initialize model parameters 
+        # Initialize model parameters
         theta = torch.zeros(n, 1)
-        wwtime = np.zeros((self.max_iter, n))
-        # Looping until max number of iterations or convergence 
+        _ = np.zeros((self.max_iter, n))
```
Comment (TheSlimvReal) [inline heat/ml/regression/lasso/lasso.py]: so i guess this can be completely removed

File: heat/core/dndarray.py [hunk 28]
```diff
@@ -2121,32 +2200,41 @@ def resplit_(self, axis=None):
...
+            self.__array = temp.clone().detach()
             self.__split = axis
 
         # entirely new split axis, need to redistribute
         else:
             _, output_shape, _ = self.comm.chunk(self.shape, axis)
             redistributed = torch.empty(output_shape, dtype=self.dtype.torch_type())
 
-            send_counts, send_displs, _ = self.comm.counts_displs_shape(self.lshape, axis)
-            recv_counts, recv_displs, _ = self.comm.counts_displs_shape(self.shape, self.split)
```
[the remark this answers is not in the document]
Comment (bhagemeier) [reply heat/core/dndarray.py]: We're currently using the default of 88 characters per line. If you think this is too short, let me know. Find the rationale [here](https://github.com/psf/black#line-length).

File: heat/core/tests/test_arithmetics.py (hunk 22 above)
Comment (bhagemeier) [reply heat/core/tests/test_arithmetics.py]: It's Python code that is part of the project, so I'm arguing yes.

File: heat/core/types.py (hunk 25 above)
Comment (bhagemeier) [reply heat/core/types.py]: This is also one of the things I noticed when first blackening the file. Unfortunately, Black is not configurable in this regard.

File: heat/ml/regression/lasso/lasso.py (hunk 27 above)
Comment (bhagemeier) [reply heat/ml/regression/lasso/lasso.py]: I would guess so. I kept the statement in order to still have the same side effects, if there are any.

File: heat/core/arithmetics.py (hunk 2 above)
Comment (bhagemeier) [reply heat/core/arithmetics.py]: It is only made a multi-line statement, because the line is too long including the comment. Putting the comment in the line before the statement, Black happily accepts it as (and transforms it back to) a single line statement.

**Comment by bhagemeier** [comment]
Guys, we need to move on with this. The PR contains changes in all areas of the code, and I want to avoid that all of you have diverged too far from it before merging.

The critical questions to resolve are:

1. quotation marks: single vs. double, where [Black](https://github.com/psf/black) enforces double quotation marks. Our contribution guide does not mandate anything in that direction. Generally, Black has a command line option to skip what they call "string normalization", which includes changing quotes;
2. placement of comments: I think I've fixed the positions. The weird positioning is almost exclusively caused by trailing comments and Black trying to stick to that.

Having Black or another tool normalize our code as part of a pre-commit hook will get rid of any trouble. Even if you prefer to put single quotes, you can rung Black manually to "fix" that if needed. We just need a decision. I propose to take it to our chat if deemed necessary.

If Black is not the right tool, the we need to decide this soon, and I will have to redo some of the work in this PR, but it's straight forward for the better part of it.

**Comment by Markus-Goetz** [comment]
Personally, I think black is doing a very fine job. We should keep it. Regarding the discussion points:

1. I can very much live with double quotation marks
2. Thanks for fixing all the places. I guess this is something that we should report on the black issue tracker (if not already done so). This is PEP-8 conform and should not be formatted in the way it is.

I have simply waited with confirming this pull request as there were still open discussion points. As soon as they are marked as resolved, I am fine with a merge.

File: heat/core/types.py (hunk 26 above)
Comment (bhagemeier) [reply heat/core/types.py]: I do agree, but Black does not give us a chance to do this.

File: heat/core/dndarray.py [hunk 29]
```diff
@@ -1586,20 +1678,6 @@ def is_distributed(self):
         """
         return self.split is not None and self.comm.is_distributed()
 
-    def item(self):
```
Comment (coquelin77) [inline heat/core/dndarray.py]: why is this deleted? is this somewhere else that im not seeing?

Comment (bhagemeier) [reply heat/core/dndarray.py]: Because it was a duplicate implementation, item() is already defined in line 173.
